logging/proto_conversions.go

```diff
@@ -0,0 +1,81 @@
+package logging
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"math"
+	"strconv"
+	"time"
+
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+	"go.viam.com/utils/protoutils"
+	"google.golang.org/protobuf/types/known/structpb"
+)
+
+// FieldToProto converts a zap.Field to a *structpb.Struct.
+func FieldToProto(field zap.Field) (*structpb.Struct, error) {
+	// Zap encodes float64s with very large int64s. Proto conversions have some
+	// loss with very large int64s. float32s are also encoded with int64s, but
+	// the int64 encodings are not large enough to cause loss in conversion.
+	// See https://pkg.go.dev/google.golang.org/protobuf@v1.32.0/types/known/structpb#NewValue.
+	//
+	// Use a hacky combination of fmt and math to store float64s as strings.
+	if field.Type == zapcore.Float64Type {
+		field.String = fmt.Sprintf("%f", math.Float64frombits(uint64(field.Integer)))
+	}
+
+	return protoutils.StructToStructPb(field)
+}
+
+// FieldKeyAndValueFromProto examines a *structpb.Struct and returns its key
+// string and native golang value.
+func FieldKeyAndValueFromProto(field *structpb.Struct) (string, any, error) {
+	var fieldValue any
+
+	fieldJSON, err := json.Marshal(field)
+	if err != nil {
+		return "", nil, err
+	}
+
+	var zf zap.Field
+	if err := json.Unmarshal(fieldJSON, &zf); err != nil {
+		return "", nil, err
+	}
+
+	// This code is modeled after zapcore.Field.AddTo:
+	// https://github.com/uber-go/zap/blob/fcf8ee58669e358bbd6460bef5c2ee7a53c0803a/zapcore/field.go#L114
+	//nolint:exhaustive
+	switch zf.Type {
+	case zapcore.BoolType:
+		fieldValue = zf.Integer == 1
+	case zapcore.DurationType:
+		fieldValue = time.Duration(zf.Integer)
+	case zapcore.Float64Type:
+		// See FieldToProto above: we encode float64s as strings to avoid loss in
+		// proto conversion.
+		if zf.String == "" {
+			return "", nil, errors.New("must encode float64s in the String field")
```

Comment:
Good question + callout.

> old modules + new rdk

Not worried about this case: old Go modules don't log fields over the wire. C++ and Python modules currently don't send any logs over the wire.

> old logs + new cli

Yeah this will break (robot has an old log with a float64 encoded only in the `Integer` field and the CLI tries to extract the key and value from that field). Still returned an error here but only warn-logged it in `robot/server/server.go`; I think we should do the same in the eventual call to this function in CLI code, @zaporter-work.